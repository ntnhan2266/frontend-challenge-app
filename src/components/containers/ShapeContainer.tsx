import React from "react";

import Rect, { RectProps } from "components/shapes/Rect";
import Circle, { CircleProps } from "components/shapes/Circle";
import { intersects, SHAPE_TYPE } from "utils/shape";

export interface ShapeContainerState {
  rectProps: RectProps;
  circleProps: CircleProps;
  isIntersected: boolean;
}

class ShapeContainer extends React.Component<any, ShapeContainerState> {
  constructor(props: any) {
    super(props);
    this.state = {
      rectProps: {
        x: 120,
        y: 120,
        width: 120,
        height: 100,
      },
      circleProps: {
        cx: 380,
        cy: 150,
        r: 30,
      },
      isIntersected: false,
    };
  }

  onDragHandler = (shapeType: SHAPE_TYPE, x: number, y: number) => {
    let newState = { ...this.state };
    switch (shapeType) {
      case SHAPE_TYPE.RECT: {
        newState = { ...newState, rectProps: { ...newState.rectProps, x, y } };
        break;
      }
      case SHAPE_TYPE.CIRCLE: {
        newState = {
          ...newState,
          circleProps: { ...newState.circleProps, cx: x, cy: y },
        };
        break;
      }
      default:
        break;
    }
    const isIntersected = intersects(newState.circleProps, newState.rectProps);
    newState = { ...newState, isIntersected };
    this.setState(newState);
  };

  render() {
    return (
      <svg width={"100vw"} height={"100vh"}>
        <Rect {...this.state.rectProps} />
        <Circle
          {...this.state.circleProps}
          onDrag={this.onDragHandler}
          isIntersected={this.state.isIntersected}
        />
      </svg>
    );
  }
}

export default ShapeContainer;
