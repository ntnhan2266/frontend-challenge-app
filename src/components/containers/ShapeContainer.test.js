import React from "react";
import { shallow } from "enzyme";
import ShapeContainer from "components/containers/ShapeContainer";
let wrapped = shallow(<ShapeContainer />);

describe("ShapeContainer", () => {
  it("should render the ShapeContainer Component correctly", () => {
    expect(wrapped).toMatchSnapshot();
  });
});
