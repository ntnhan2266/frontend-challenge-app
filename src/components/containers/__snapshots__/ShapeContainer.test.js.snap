// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShapeContainer should render the ShapeContainer Component correctly 1`] = `
<svg
  height="100vh"
  width="100vw"
>
  <Rect
    height={100}
    width={120}
    x={120}
    y={120}
  />
  <_temp
    cx={380}
    cy={150}
    isIntersected={false}
    onDrag={[Function]}
    r={30}
  />
</svg>
`;
